lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py

```diff
@@ -0,0 +1,146 @@
+from lldbsuite.test import lldbplatformutil
+import lldbgdbserverutils
+import lldbdap_testcase
+import os
+import shutil
+import subprocess
+import tempfile
+import threading
+import time
+import sys
+
+
+class TestDAP_attachByPortNum(lldbdap_testcase.DAPTestCaseBase):
+    def runTargetProgramOnPort(self, port=None, program=None):
+        server_tool = None
+        if lldbplatformutil.getPlatform() == "linux":
+            server_tool = lldbgdbserverutils.get_lldb_server_exe()
+            if server_tool is None:
+                self.dap_server.request_disconnect(terminateDebuggee=True)
+                self.assertIsNotNone(server_tool, "lldb-server not found.")
+            server_tool += " g localhost:" + port + " "
+        elif lldbplatformutil.getPlatform() == "macosx":
+            server_tool = lldbgdbserverutils.get_debugserver_exe()
+            if server_tool is None:
+                self.dap_server.request_disconnect(terminateDebuggee=True)
+                self.assertIsNotNone(server_tool, "debugserver not found.")
+            server_tool += " --listen localhost:" + port + " "
+
+        self.process = subprocess.Popen(
+            [server_tool + program],
+            shell=True,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+
+        return self.process
+
+    def set_and_hit_breakpoint(self, continueToExit=True):
+        source = "main.c"
+        main_source_path = os.path.join(os.getcwd(), source)
+        breakpoint1_line = line_number(main_source_path, "// breakpoint 1")
+        lines = [breakpoint1_line]
+        # Set breakpoint in the thread function so we can step the threads
+        breakpoint_ids = self.set_source_breakpoints(main_source_path, lines)
+        self.assertEqual(
+            len(breakpoint_ids), len(lines), "expect correct number of breakpoints"
+        )
+        self.continue_to_breakpoints(breakpoint_ids)
+        if continueToExit:
+            self.continue_to_exit()
+
+    @skipIfWindows
+    @skipIfNetBSD  # Hangs on NetBSD as well
+    @skipIfRemote
+    def test_by_port(self):
+        """
+        Tests attaching to a process by port.
+        """
+        self.build_and_create_debug_adaptor()
+        program = self.getBuildArtifact("a.out")
+
+        port = "2345"
+        self.process = self.runTargetProgramOnPort(port=port, program=program)
+        pid = self.process.pid
+        response = self.attach(program=program, port=int(port), sourceInitFile=True)
+        self.set_and_hit_breakpoint(continueToExit=True)
+        self.process.kill()
+
+    @skipIfWindows
+    @skipIfNetBSD  # Hangs on NetBSD as well
+    @skipIfRemote
+    def test_by_port_and_pid(self):
+        """
+        Tests attaching to a process by process ID and port number.
+        """
+        self.build_and_create_debug_adaptor()
+        program = self.getBuildArtifact("a.out")
+
+        port = "2345"
```

Comment:
use a different port for each test, just in case the test runner executes them simultaneously

Reply:
That's definitely not enough, and it's guaranteed to make the test flaky and/or environment sensitive. Both lldb-server and gdb support race-free port allocations by choosing the ports themselves letting the caller know of the actual chosen port.

Writing this logic in a way that works across server implementations and operating systems is not easy, but we already have the code (several in fact, but I'd recommend reusing the implementation in  `test/API/tools/lldb-server/commandline/TestGdbRemoteConnection.py`).

As an alternative, since what you really just want to test is that the connection is made, you could consider using the mock gdb server implementation in `test/API/functionalities/gdb_remote_client/`. That way, it will be you (the python code) who's opening the ports, and there will again be no races.